components/led_strip/led_strip.c

```diff
@@ -204,53 +211,148 @@ static void led_strip_fill_rmt_items_apa106(struct led_color_t *led_strip_buf, r
     }
 }
 
-static void led_strip_task(void *arg)
+void apa102_init(struct led_strip_t *led_strip)
 {
-    struct led_strip_t *led_strip = (struct led_strip_t *)arg;
-    led_fill_rmt_items_fn led_make_waveform = NULL;
+    ESP_LOGI(TAG, "apa102 init clk:%d data:%d", led_strip->clk, led_strip->gpio);
+    gpio_config_t d_gpio_config = {
+        .mode = GPIO_MODE_OUTPUT,
+        .pin_bit_mask = 1ULL << led_strip->gpio
+    };
+    ESP_ERROR_CHECK(gpio_config(&d_gpio_config));
+    gpio_set_level(led_strip->gpio, 0);
+    gpio_config_t c_gpio_config = {
+        .mode = GPIO_MODE_OUTPUT,
+        .pin_bit_mask = 1ULL << led_strip->clk
+    };
+    ESP_ERROR_CHECK(gpio_config(&c_gpio_config));
+    gpio_set_level(led_strip->clk, 0);
+}
+
+static void apa102_transfer(uint8_t dataPin, uint8_t clockPin, uint8_t b)
```

Comment:
Could you make that a loop?

Reply:
Added a macro

Reply:
Macros tend to bloat the generated assembly code and there's really no optimization, whereas loops are more efficient because the same assembly is executed. I'm actually revisiting some older coffee of mine too squeeze out as much efficiency as possible because we're short in flash

Reply:
> Macros tend to bloat the generated assembly code and there's really no optimization, whereas loops are more efficient 

The macro is a loop.  I can create a new PR with the loop directly in code, but it does not effect the bin size.

| Partition | Original code | Current Loop Macro | Loop Macro (no N) | Loop in code |
| --------- | --------- | --------- | --------- | --------- |
| recovery.bin | 1310352 | 1310320 | 1310320 | 1310320 |
| squeezlelite.bin | 2647616 | 2647584 | 2647584 | 2647584 |